lib/floe/workflow/state.rb

```diff
@@ -45,18 +45,19 @@ def status
       def run!(input)
         logger.info("Running state: [#{name}] with input [#{input}]")
 
-        input = input_path.value(context, input)
-
-        output, next_state = block_given? ? yield(input) : input
-        next_state ||= workflow.states_by_name[payload["Next"]] unless end?
-
-        output ||= input
-        output   = output_path&.value(context, output)
+        output     = execute!(input)
+        next_state = workflow.states_by_name[@next] unless end?
```

Comment:
I like the stateless feel of `output = execute!(input)`. The feel that it is reading from the input and returning the output, and there aren't really any other side effects. But maybe `context` breaks this and it is an illusion. The `@next` is then jarring and shatters my nice stateless command pattern dreams.

I think of this like mouse routines where you have to convert the input to the local coordinate system and back to the global system.
`local` and `global` may be the wrong paradigm here, but this is what I'm thinking.

translate may be a better phrase, but I like the idea of the method name saying "I'm converting outside stuff to local stuff" and "I'm converting local stuff to outside stuff"


```ruby
  def localize_input(input)
    input_path ? input_path.value(context, input) : input
  end

  def globalize_next(next)
    workflow.states_by_name[next] if next && !end?
  end

  def globalize_output(output)
    output_path ? output_path.value(context, output) : output
  end

  # unsure if this is necessary, but feels like a parent thing
  def end?
    true
  end

  def run(input)
    next, output     = execute!(localize_input(input))
    [globalize_next(next), globalize_output(output)]
  end

```

It seems like `next` is pretty much hard coded except for `choice` (obvious why this is the case) and task (not as obvious but don't know what a task is and haven't been thinking too hard on it). So I can see why you wouldn't want to return from every state transition.

`end?` seems like it is overridden to be the same thing for all, so maybe moving that call into the states and removing that call from `globalize_next`/`State` may make sense.

Reply:
I feel like `run` and parent `State` should know about the other workflows, and `execute` is only worried about the nuances of the particular state node.

But this may be naïve since transitions like `choice` are all about choosing from the workflow provided values.